ArduCopter/control_guided.cpp

```diff
@@ -586,6 +587,45 @@ void Copter::guided_angle_control_run()
     pos_control.update_z_controller();
 }
 
+// helper function to update position controller's desired velocity while respecting acceleration limits
+void Copter::guided_set_desired_velocity_with_accel_limits(const Vector3f& vel_des)
+{
+    // get current desired velocity
+    Vector3f curr_vel_des = pos_control.get_desired_velocity();
+
+    // exit immediately if already equal
+    if (curr_vel_des == vel_des) {
+        return;
+    }
+
+    // get change in desired velocity
+    Vector3f vel_delta = vel_des - curr_vel_des;
+
+    // limit xy change
+    float vel_delta_xy = safe_sqrt(sq(vel_delta.x)+sq(vel_delta.y));
+    float vel_delta_xy_max = G_Dt * pos_control.get_accel_xy();
+    float ratio_xy = 1.0f;
+    if (!is_zero(vel_delta_xy) && (vel_delta_xy > vel_delta_xy_max)) {
+        ratio_xy = vel_delta_xy_max / vel_delta_xy;
+    }
+
+    // limit z change
+    float vel_delta_z = fabsf(vel_delta.z);
+    float vel_delta_z_max = G_Dt * pos_control.get_accel_z();
+    float ratio_z = 1.0f;
+    if (!is_zero(vel_delta_z) && (vel_delta_z > vel_delta_z_max)) {
+        ratio_z = vel_delta_z_max / vel_delta_z;
+    }
+
+    // new target
+    curr_vel_des.x += (vel_delta.x * ratio_xy);
+    curr_vel_des.y += (vel_delta.y * ratio_xy);
+    curr_vel_des.z += (vel_delta.z * ratio_z);
```

Comment:
Is this equivalent to:

`curr_vel_des.z += MIN(vel_delta_z, vel_delta_z_max)`

That would allow you to eliminate a division, multiplication, `ratio_z`, and the above if statement.